src/hyp_map.rs

```diff
@@ -260,11 +374,33 @@ impl HypMap {
             .unwrap();
         let sv48: FirstStagePageTable<Sv48> =
             FirstStagePageTable::new(root_page).expect("creating first sv48");
-        // Map regions shared across all pagetables.
+
+        // Map all the regions in the memory map that the hypervisor could need.
         for r in &self.shared_regions {
-            r.map(&sv48, &mut || {
-                hyp_mem.take_pages_for_hyp_state(1).into_iter().next()
-            });
+            if r.variable_mapped {
+                // split region
+                // r_start and r_end are only used for asserts
+                let r_start = r.paddr.bits();
+                let r_end = r_start + r.page_size() as u64 * r.page_count as u64;
+                if let Some(parts) = r.partition() {
+                    for i in parts {
+                        let i_start = i.paddr.bits();
+                        let i_end = i_start + i.page_size() as u64 * i.page_count as u64;
+                        assert!(i_start >= r_start && i_start < r_end);
+                        assert!(i_end > r_start && i_end <= r_end);
+                        i.map(&sv48, &mut || {
+                            hyp_mem
+                                .take_pages_for_hyp_state_with_size(1, i.page_size())
+                                .into_iter()
+                                .next()
+                        })
```

Comment:
The closure passed to `map()` is meant to retrieve a free page-table page, which are always 4kB.